### PR TITLE
build: enable pyright

### PR DESCRIPTION
**Pull request by lengau**

### CODE REVIEW

File: charmcraft/commands/extensions.py [hunk 1]
```diff
@@ -47,17 +47,18 @@ def run(self, parsed_args: argparse.Namespace):
         """Print the list of available extensions and their bases."""
         extension_presentation: dict[str, ExtensionModel] = {}
 
-        for extension_name in extensions.registry.get_extension_names():
-            extension_class = extensions.registry.get_extension_class(extension_name)
+        for extension_name in extensions.get_extension_names():
```
Comment (syu-w) [inline charmcraft/commands/extensions.py:50]: They are the same thing?

**Review by syu-w — APPROVED** [review]

File: charmcraft/commands/extensions.py (hunk 1 above)
Comment (lengau) [reply charmcraft/commands/extensions.py:50]: Yep, but because `extensions` doesn't export `registry`, pyright complains.

**Review by mr-cal — APPROVED** [review]
Nice work, has charmcraft never been pyright-compliant?

**Comment by lengau** [comment]
@mr-cal nope! In fact I only even introduced mypy a few months ago.

It still requires some work though. I've bumped a lot of errors down to warnings, but...

```
$ pyright --pythonpath .venv/bin/python | grep -o -E '\([A-Za-z]+\)' | sort | uniq -c | sort -h
      1 (reportCallIssue)
      2 (reportUnknownLambdaType)
      4 (reportReturnType)
      4 (reportUnsupportedDunderAll)
      6 (reportUnnecessaryIsInstance)
      7 (reportIncompatibleMethodOverride)
     11 (reportGeneralTypeIssues)
     13 (reportIncompatibleVariableOverride)
     22 (reportUntypedFunctionDecorator)
     24 (reportArgumentType)
     26 (reportMissingTypeArgument)
     26 (reportOptionalMemberAccess)
     35 (reportAttributeAccessIssue)
    136 (reportUnknownArgumentType)
    173 (reportUnknownVariableType)
    249 (reportMissingParameterType)
    262 (reportUnknownMemberType)
    378 (reportUnknownParameterType)
```

**Comment by mr-cal** [comment]
Wow, well nice work with adding in linting in a gradual way!

**Pull request merged**